src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,240 @@
+
+    try:
+        response = requests.get(API_URL, params)
+    
+        if response.status_code == 200:
+            print("API request has been successful")
+            return response.json()
+        
+        elif response.status_code == 429:
+            print("Rate limit exceeded. Please try again later.")
+            return None
+        else:
+            handle_api_error(response) 
+            return None
+        
+    except:
+        print(f"API Request failed: {requests.RequestException}")
+        return None
+    
+
+""" Given a city, defines the parameters and calls the API. 
+If the call is successful, it returns the corresponding daily dataframe; otherwise, it tries reading the
+corresponding csv file."""
+def get_data_meteo_api(city):
+
+    latitude = COORDINATES[city]['latitude']
+    longitude = COORDINATES[city]['longitude']
+
+    params = {
+	"latitude": latitude,
+	"longitude": longitude,
+	"start_date": "1950-01-01",   # from 1950 to 2050
+	"end_date": "2050-12-31",
+	"models": "EC_Earth3P_HR", # One of the models with all weather variables available
+	"daily": ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"] # Data we want to extract
+    }
+
+    # Call the API
+    responses = call_API(params)
+
+    file_path = os.path.join(OUTPUT_DIR, f'{city}_daily_data.csv')
+
+    if responses:     
+        daily_json = get_daily_data(responses, city)
+        # Transform to dataframe
+        daily_df = pd.DataFrame(daily_json)
+    
+        # Save it in a csv only if there is not already one
+        if not os.path.exists(file_path):
+            daily_df.to_csv(file_path, index=False)
+        
+        return daily_df
+    
+    else:
+        try:
+            print("Failed to get API data. Loading data from CSV.")
+            return pd.read_csv(file_path)
+        
+        except FileNotFoundError:
+            print("No csv file found.")
+            return None
+    
+
+""" Given the response from the API and the city, gets the returned time series and saves them in a dataframe. 
+The dataframe is then saved in a csv ({city}_daily_data.csv). """
+def get_daily_data(response, city):
+    daily_data = response.get('data', {})
+    
+    date = daily_data.get('time', [])
+    daily_temperature_2m_mean = daily_data.get('temperature_2m_mean', [])
+    daily_precipitation_sum = daily_data.get('precipitation_sum', [])
+    daily_soil_moisture_0_to_10cm_mean = daily_data.get('soil_moisture_0_to_10cm_mean', [])
+    
+    daily_df = daily_df = pd.DataFrame({
+        "date": date,
+        "temperature_2m_mean": daily_temperature_2m_mean,
+        "precipitation_sum": daily_precipitation_sum,
+        "soil_moisture_0_to_10cm_mean": daily_soil_moisture_0_to_10cm_mean
+    })
+
+    return daily_df
+    
+
+""" Given a daily dataframe, transforms it into mothly data. """
+def reduce_temporal_resolution(daily_df, city):
+     
+    daily_df['date'] = pd.to_datetime(daily_df['date'])
```

Comment:
If this is a date consider to cast datetime to .date to avoid problems if somehow there are different hour values